### PR TITLE
Lazy load commands

### PR DESCRIPTION
**Pull request by tamorim**
This PR adds lazy load capability through the `module` property.

### CODE REVIEW

File: src/finder.js [hunk 1]
```diff
@@ -1,5 +1,22 @@
...
+  reject,
+  reduce,
+  propEq,
+  values,
+  length,
+  compose,
+  groupBy,
+  flatten,
+  find as rfind,
+} from 'ramda'
```
Comment (vibegui) [inline src/finder.js:19]: Nice :joy:

File: src/finder.js [hunk 2]
```diff
@@ -32,7 +49,15 @@ export function isCommand (node) {
 }
 
 export function isNamespace (node) {
-  return type(node) === 'Object' && any(isCommand, values(node)) && node
+  return type(node) === 'Object' && any(v => isCommand(v) || isModule(v), values(node)) && node
```
Comment (vibegui) [inline src/finder.js:52]: I feel that `v => isCommand(v) || isModule(v)` is _asking_ to be _ramda'ed_, but I couldn't find a nice way to do it. 😁

File: src/finder.js [hunk 3]
```diff
@@ -32,7 +49,15 @@ export function isCommand (node) {
...
+}
+
+export function isModule (node) {
+  return !isNil(node) && type(node.module) === 'String' && node
+}
+
+export function loadModule (node) {
+  return require(node.module).default
```
Comment (vibegui) [inline src/finder.js]: I'd add compatibility for non-babel/ES6 modules (e.g `return require(node.module).default || require(node.module)`)

File: src/finder.js [hunk 4]
```diff
@@ -55,7 +80,11 @@ export function find (node, args, raw, minimist) {
   }
 
   const [head, ...tail] = args
-  const next = findNext(head, node)
+  let next = findNext(head, node)
+
+  if (isModule(next)) {
```
Comment (vibegui) [inline src/finder.js]: Why not encapsulate this inside `findNext`?

File: src/finder.js [hunk 5]
```diff
@@ -76,5 +105,6 @@ export function find (node, args, raw, minimist) {
 }
 
 export function run ({command, args}) {
-  return command.handler.apply(this, args)
+  const cmd = isModule(command) ? loadModule(command) : command
```
Comment (vibegui) [inline src/finder.js]: When have you felt that the client of `run` would send an _unrequired_ module?

File: src/helper.js [hunk 6]
```diff
@@ -22,6 +37,27 @@ ${map(formatOption, rootOptions.options).join('\n')}
 `
 }
 
+export function loadTree (tree) {
+  return mapObjIndexed(n => isModule(n) ? loadModule(n) : n)(tree)
```
Comment (vibegui) [inline src/helper.js]: Unnecessary double-call, use parens and comma instead? :)

File: src/helper.js [hunk 7]
```diff
@@ -22,6 +37,27 @@ ${map(formatOption, rootOptions.options).join('\n')}
...
+  return groupBy(([, node]) =>
+    isOptions(node) ? 'rootOptions'
+    : isCommand(node) && !isNamespace(node) ? 'root'
+    : isNamespace(node) ? 'ns'
+    : '_'
+  )(toPairs(tree))
+}
+
+export function pairsToTree (pairs) {
+  return mapObjIndexed(n => fromPairs(n))(pairs)
```
Comment (vibegui) [inline src/helper.js]: `n => fromPairs(n)` === `fromPairs`? Also, double call!

File: src/helper.js [hunk 8]
```diff
@@ -22,6 +37,27 @@ ${map(formatOption, rootOptions.options).join('\n')}
...
+  return mapObjIndexed(n => isModule(n) ? loadModule(n) : n)(tree)
+}
+
+export function groupTree (tree) {
+  return groupBy(([, node]) =>
+    isOptions(node) ? 'rootOptions'
+    : isCommand(node) && !isNamespace(node) ? 'root'
+    : isNamespace(node) ? 'ns'
+    : '_'
+  )(toPairs(tree))
```
Comment (vibegui) [inline src/helper.js]: Can you expand (maybe also add comments) on the necessity of converting to pairs?

**Review by vibegui — CHANGES_REQUESTED** [review]
Awesome first iteration!!!

**Comment by tamorim** [comment]
@firstdoit Done!

**Comment by vibegui** [comment]
For future reference, avoid push force after a review, because GitHub apparently gets lost :P

**Comment by tamorim** [comment]
@firstdoit Noted.

**Comment by tamorim** [comment]
@firstdoit Latest optimizations on `helper` are done! Thank you for the extra brain on the refactor :+1:

**Comment by vibegui** [comment]
Thanks for letting me!

**Comment by vibegui** [comment]
maravilhoso

**Review by vibegui — APPROVED** [review]
LGTM!

**Pull request merged**